### ISSUE TITLE
Apply linting against the specs in the source code

### ISSUE THREAD

**Issue by tonyhallett** [nestjs/nest#3491]
## Is your feature request related to a problem? Please describe.

Some of the tests of the source suffer from one of two problems.

a) Expectation inside catch handlers or within promise catch methods.

I have cleared a few of these locally and will make a pull request shortly but there still remains 22 unit tests and 5 integration tests that fall in to this category.  Of the ones that I changed there was at least one where the test had not been set up correctly and no error was thrown when it was expected to.

b) Expectations that are incomplete and do not execute - [e.g](https://github.com/nestjs/nest/blob/36d5a81439144550b2a6f9d1f3ac79f32a4656fe/packages/websockets/test/container.spec.ts#L24).

There were 4 of the type expect('this').to.be.someMethod without the ().  There were 5 of the type expect(spy.calledWith('someValue')) that should have ended with .to.be.true;
I have cleared all of these locally and will make a pull request.  Again I found tests that failed once the expectation had been corrected.

## Describe the solution you'd like
Run tslint against the tests in the integration and packages folder with custom tslint rules that address each issue. 

I have already created the rules in the package expect-ts-lint ( [npm](https://www.npmjs.com/package/expect-tslint) and [github](https://github.com/tonyhallett/expect-tslint) ).  

It has been a while since I was proficient with the typescript compiler API so the rules could possibly be made more efficient or cover cases that I have not considered.  See [chai-terminated](https://github.com/tonyhallett/expect-tslint/blob/master/test/rules/chai-terminated-rule/additional-terminals/test.ts.lint) and [no-expect-in-catch](https://github.com/tonyhallett/expect-tslint/blob/master/test/rules/no-expect-in-catch/aliases/test.ts.lint) for code that will error when tslint is run.

If you would like to see them in operation without having to set up within your nest repo then you can clone [lint-nest-specs](https://github.com/tonyhallett/lint-nest-specs) and run the script lint-nest-spec.

## Teachability, Documentation, Adoption, Migration Strategy
As this is for contributors and linting is well known and the error messages are clear there should be no need.

## What is the motivation / use case for changing the behavior?
Tests can be created that appear to pass but the expectation is never run.

**Changes required if use expect-tslint**
a) Install the package
b) Create a separate tslint.json for the two rules.
c) Create a script entry in package.json - lint-specs
d) **Enforce this script to be run**
**Currently it is possible to locally commit typescript that would fail the linting process.**  I am assuming that CircleCI is used and it catches tslint errors ? Even if it is, shouldn't the lint-staged entry in package.json be changed to run tslint ?
```
"lint-staged": {
    "packages/**/*.{ts,json}": [
      "npm run format",
      "git add"
    ]
  },
```
e) Correct all of the tests that expect inside catch.

**Comment by kamilmysliwiec** [comment]
Thanks for your feedback!

- a) Expectation inside catch handlers or within promise catch methods.
- b) Expectations that are incomplete and do not execute - e.g.

PRs to fix both would be very welcome if you're willing to create them

> Currently it is possible to locally commit typescript that would fail the linting process. I am assuming that CircleCI is used and it catches tslint errors ? Even if it is, shouldn't the lint-staged entry in package.json be changed to run tslint ?

We are using CircleCI for that, but I agree with you and have already added `tslint` to `lint-staged` in order to run it with every commit.

As for custom tslint rules, I don't think that we necessarily need them. These issues that you've found are very likely one-time fixes. It should be fine to just fix them now and pay more attention to such cases in the future (during code reviews).

**Comment by tonyhallett** [comment]
Agreed they are one time fixes and if more attention is paid in code reviews then great.  Although if you used these rules then such errors would never be committed and it would be one less task for code reviews.

I will create pull requests for the errors that I have already corrected but I do not have the time to resolve the remaining.  They may be resolvable with a simple refactoring but from what I have already found there will be some that take a fair bit of work to set up the test appropriately.

If you are interested these are the remaining tests with errors.

packages/common/test/decorators/exception-filters.decorator.spec.ts:38:7 
ERROR: 38:7  no-expect-in-catch  Expect call in catch handler

packages/common/test/decorators/use-guards.decorator.spec.ts:44:7 
ERROR: 44:7  no-expect-in-catch  Expect call in catch handler

packages/common/test/decorators/use-pipes.decorator.spec.ts:35:7 
ERROR: 35:7  no-expect-in-catch  Expect call in catch handler

packages/microservices/test/client/client-grpc.spec.ts:206:11 
ERROR: 206:11  no-expect-in-catch  Expect call in catch handler 
ERROR: 249:39  no-expect-in-catch  Expect call in promise catch 
ERROR: 256:9   no-expect-in-catch  Expect call in promise catch

packages/microservices/test/client/client-kafka.spec.ts:469:9 
ERROR: 469:9  no-expect-in-catch  Expect call in promise catch

packages/microservices/test/client/client-mqtt.spec.ts:307:9 
ERROR: 307:9  no-expect-in-catch  Expect call in promise catch

packages/microservices/test/client/client-nats.spec.ts:268:9 
ERROR: 268:9  no-expect-in-catch  Expect call in promise catch

packages/microservices/test/client/client-redis.spec.ts:338:9 
ERROR: 338:9  no-expect-in-catch  Expect call in promise catch

packages/microservices/test/client/client-rmq.spec.ts:344:9 
ERROR: 344:9  no-expect-in-catch  Expect call in promise catch

packages/microservices/test/context/rpc-context-creator.spec.ts:132:13 
ERROR: 132:13  no-expect-in-catch  Expect call in promise catch 
ERROR: 153:33  no-expect-in-catch  Expect call in promise catch

packages/microservices/test/json-socket/message-parsing.spec.ts:119:11 
ERROR: 119:11  no-expect-in-catch  Expect call in catch handler 
ERROR: 157:11  no-expect-in-catch  Expect call in catch handler

packages/microservices/test/server/server-grpc.spec.ts:64:11 
ERROR: 64:11  no-expect-in-catch  Expect call in catch handler

packages/platform-express/test/multer/interceptors/any-files.interceptor.spec.ts:41:18 
ERROR: 41:18  no-expect-in-catch  Expect call in promise catch

packages/platform-express/test/multer/interceptors/file-fields.interceptor.spec.ts:60:18 
ERROR: 60:18  no-expect-in-catch  Expect call in promise catch

packages/platform-express/test/multer/interceptors/file.interceptor.spec.ts:42:18 
ERROR: 42:18  no-expect-in-catch  Expect call in promise catch

packages/platform-express/test/multer/interceptors/files.interceptor.spec.ts:44:18 
ERROR: 44:18  no-expect-in-catch  Expect call in promise catch

packages/websockets/test/context/ws-context-creator.spec.ts:127:13 
ERROR: 127:13  no-expect-in-catch  Expect call in promise catch 
ERROR: 148:33  no-expect-in-catch  Expect call in promise catch

integration/injector/e2e/injector.spec.ts:22:9 
ERROR: 22:9  no-expect-in-catch  expect call in catch handler 
ERROR: 35:9  no-expect-in-catch  expect call in catch handler

integration/injector/e2e/scoped-instances.spec.ts:51:7 
ERROR: 51:7  no-expect-in-catch  expect call in catch handler

integration/nest-application/get-url/e2e/express.spec.ts:45:7 
ERROR: 45:7  no-expect-in-catch  expect call in catch handler

integration/nest-application/get-url/e2e/fastify.spec.ts:38:7 
ERROR: 38:7  no-expect-in-catch  expect call in catch handler
